lib/task.ml

```diff
@@ -32,7 +40,15 @@ let setup_pool ~num_additional_domains =
         worker ()
   in
   let domains = Array.init num_additional_domains (fun _ -> Domain.spawn worker) in
-  {domains; task_chan}
+  let p = {domains; task_chan} in
+  let _ = match name with
+    | Some x ->
+      Mutex.lock named_pool_mutex;
+      Hashtbl.add named_pools x p;
```

Comment:
Meta-comment: I started thinking about the possibility of exceptions from `Hashtbl.add`. While `Hashtbl.add` does not raise any synchronous exceptions, there is of course the possibility of asynchronous exceptions. For example, `add` performs an allocation, which triggers a GC, which runs a finaliser which raises an exception. Because of the exception, the mutex would not be unlocked. 

It would be useful to have a `with_mutex` primitive in `Mutex` module:

```ocaml
let with_mutex m f v =
  Mutex.lock m;
  begin try f v with e -> 
    Mutex.unlock m;
    raise e
  end;
  Mutex.unlock m
```

which will guard against any exceptional behaviour. CC @ctk21 @sadiqj.

Reply:
that seems not unreasonable to me